global/scripts/get_ROI_from_txt.m

```diff
@@ -0,0 +1,32 @@
+function first_column_values = get_ROI_from_txt(filename)
```

Comment:
Given that this reads only a list of numbers from a text file, please consider not calling it `get_ROI...`.  Maybe `get_label_keys_from_txt`.

Again, though, a lot of this matlab complexity can be avoided if the bash script just does this a few times before calling matlab:
```
wb_command -cifti-create-dense-from-template <grayordinates> <fulldlabel> -cifti <dlabel>
wb_command -cifti-label-import <fulldlabel> <txt> <selectedkeys>
wb_command -cifti-math 'x > 0' <roi> -var x <selectedkeys>
```

Reply:
Perhaps I have been a bit too obtuse, let me rephrase: Unless there is a good reason not to, PLEASE remove the matlab code for labels.txt format reading, _extract_data, etc index-fixing code, and instead make the ROIs in bash (as temporary files, if appropriate, using `tempfiles_create`) and pass those into matlab instead of the text files.  This will reduce the amount of code to look over in the PR, and more of it will be focused on the new features, rather than dealing with file format issues.

It also looks like you are using a group-consensus dlabel file, but this code is run per-subject, so we could also consider trying not to do this repeatedly for each subject.

Reply:
Hi Tim, can you provide me with a more detailed example to achieve this via workbench command? For example, for the visual roi.

Reply:
This will do it, with the final temporary filename in the `VisualROItemp` variable:
```
#you already have this source line in the appropriate script, but for reference:
source "$HCPPIPEDIR"/global/scripts.tempfiles.shlib "$@"
...
tempfiles_create MMPROIs_temp_XXXXXX.dlabel.nii roitemp
wb_command -cifti-create-dense-from-template "$HCPPIPDEDIR"/global/templates/91282_Greyordinates.dscalar.nii "$roitemp" -cifti <original_MMP_59k>

tempfiles_create VisualROI_temp_XXXXXX.dscalar.nii VisualROItemp
tempfiles_add "$VisualROItemp"_labels.dlabel.nii
wb_command -cifti-label-import "$roitemp" "$HCPPIPDEDIR"/global/config/Visual.ROI.txt "$VisualROItemp"_labels.dlabel.nii
wb_command -cifti-math 'x > 0' "$VisualROItemp" -var x "$VisualROItemp"_labels.dlabel.nii
```
You may already know this, but you don't need to manually delete the temporaries, tempfiles.shlib does that for you when the script exits (or errors), as long as you don't background a function or otherwise subshell things.

Reply:
Hi Tim, I found when I load the file "/tmp/VisualROI_temp_.dlabel.nii" into matlab and compare with the raw `VisualROI`, they are giving me different vectors. Can you please check why it is the case? I have attached the reproducible code blocks. 

```matlab
CorticalParcellationFile='/media/myelin/alex/pipelines/pipeline_reclean_v2/HCPpipelines/global/templates/tICA/Q1-Q6_RelatedValidation210.CorticalAreas_dil_Final_Final_Areas_Group_Colors.32k_fs_LR.dlabel.nii';
CorticalParcellation=ciftiopen(CorticalParcellationFile,'wb_command');
VisualROI=single(zeros(length(CorticalParcellation.cdata),1));
VisualAreas=[1 2 3 4 5 6 7 13 16 17 18 19 20 21 22 23 48 49 142 152 153 154 156 158 159 160 163 181 182 183 184 185 186 187 193 196 197 198 199 200 201 202 203 228 229 322 332 333 334 336 338 339 340 343];
FullGreyordinateLength=91282;
for i=VisualAreas
VisualROI(CorticalParcellation.cdata==i)=1;
end
VisualROI=[VisualROI' single(zeros(1,FullGreyordinateLength-length(VisualROI)))]';
```

```bash
CorticalParcellationFile="/media/myelin/alex/pipelines/pipeline_reclean_v2/HCPpipelines/global/templates/tICA/Q1-Q6_RelatedValidation210.CorticalAreas_dil_Final_Final_Areas_Group_Colors.32k_fs_LR.dlabel.nii"
wb_command -cifti-create-dense-from-template /media/myelin/alex/pipelines/pipeline_reclean_v2/HCPpipelines/global/templates/91282_Greyordinates/91282_Greyordinates.dscalar.nii /tmp/MMPROI_tmp.dlabel.nii -cifti "$CorticalParcellationFile"

wb_command -cifti-label-import /tmp/MMPROI_tmp.dscalar.nii /media/myelin/alex/pipelines/pipeline_reclean_v2/HCPpipelines/global/config/Visual.ROI.txt /tmp/MMPROI_tmp.dscalar.nii_labels.dlabel.nii
wb_command -cifti-math 'x > 0' /tmp/VisualROI_temp_.dlabel.nii -var x /tmp/MMPROI_tmp.dscalar.nii_labels.dlabel.nii
```

Reply:
Save the matlab result to cifti (cifti_write_from_template should help), and compare them in wb_view.

Also, not having random characters after the `_` suggests that tempfiles_create didn't work properly...

Reply:
It seems that this file `/tmp/VisualROI_temp_.dlabel.nii` has the whole cortex to be set as V1... That's an issue. I intentionally didn't include random character to let it easy for you to reproduce the issue.

Reply:
You somehow went from /tmp/MMPROI_tmp.dlabel.nii to /tmp/MMPROI_tmp.dscalar.nii in your second block of code.  Is that an editing error, or in your original?  The output of the -cifti-math should also be .dscalar.nii type, not .dlabel.nii.

Sorry, I missed an important bit, you should include "-discard-others" in the -cifti-label-import command.

Reply:
Actually, I may not have special-cased label maps in -cifti-math to replace them with scalar maps, which is a bit unfortunate, so it may still tell you to use dlabel for the output.

Reply:
Did you get the wb_command way working for you?  If not, can you at least give this function a better name?

Reply:
As for `SubRegions`, where you want to keep them separate, you can just do -cifti-all-labels-to-rois after the import, which will give you a set of binary maps.  wmparc should be able to be changed to the same method as the MMP, using a different file.

Reply:
> Does it also work for non-gray regions?

The principle works for any label data, you'd just need to start with something like the wmparc file instead of the MMP (-volume-import-label and volume versions of the other commands exist, and it appears you even use some of them already).